### DISCUSSION TITLE
How to use IDENTITY_INSERT with SqlServer?

### DISCUSSION THREAD

**Discussion by CapitaineToinon** [prisma/prisma#15291] [Q&A] [answered]
Hey there ! 👋  I've been trying to populate an SqlServer database with data for e2e testing purpuses using prisma but I've been unable to set the INSERT_IDENTITY variable in order to insert data with hard coded primary keys. 

[It's a pretty normal feature of SqlServer](https://docs.microsoft.com/en-us/sql/t-sql/statements/set-identity-insert-transact-sql?view=sql-server-ver16) but I just cannot get it to work whatever I try.

I tried with `$executeRaw`

```js
await prisma.$executeRaw`SET IDENTITY_INSERT Customer ON`;
await prisma.customer.create({
  data: { id: 5 /* ... other data here */ },
});
await prisma.$executeRaw`SET IDENTITY_INSERT Customer OFF`;
```

I tried with `$executeRaw` and `$transaction`

```js
await prisma.$transaction([
  prisma.$executeRaw`SET IDENTITY_INSERT Customer ON`,
  prisma.customer.create({
    data: { id: 5 /* ... other data here */ },
  }),
  prisma.$executeRaw`SET IDENTITY_INSERT Customer OFF`,
]);
```

The generated SQL seems to be okay (until the ROLLBACK ofc):

```bash
prisma:query BEGIN TRAN
prisma:query SET IDENTITY_INSERT Customer ON
prisma:query DECLARE @generated_keys table([id] BIGINT) INSERT INTO [dbo].[Customer] ([id],[<...other fields>]) OUTPUT [Inserted].[id] INTO @generated_keys VALUES (@P1,@P2,@P3,@P4,@P5,@P6,@P7,@P8,@P9,@P10,@P11,@P12,@P13,@P14,@P15,@P16,@P17,@P18,@P19,@P20,@P21,@P22,@P23,@P24,@P25,@P26,@P27,@P28,@P29,@P30,@P31,@P32,@P33,@P34,@P35,@P36,@P37,@P38,@P39,@P40,@P41,@P42,@P43,@P44,@P45,@P46) SELECT [t].[id] FROM @generated_keys AS g INNER JOIN [dbo].[Customer] AS [t] ON [t].[id] = [g].[id] WHERE @@ROWCOUNT > 0 /* traceparent=00-00-00-00 */
prisma:query ROLLBACK
```

But I still get an error telling me it's impossible to insert values without INDENTITY_INSERT ON.

```
Error: 
Invalid `prisma.$executeRaw()` invocation:


Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(TokenError { code: 544, state: 1, class: 16, message: "Cannot insert explicit value for identity column in table 'Customer' when IDENTITY_INSERT is set to OFF.", server: "a6d5592dc4ee", procedure: "", line: 1 }) })
    at RequestHandler.handleRequestError (/Users/antoine/Documents/www/karpeo/frontend/node_modules/@prisma/client/runtime/index.js:29913:13)
    at RequestHandler.request (/Users/antoine/Documents/www/karpeo/frontend/node_modules/@prisma/client/runtime/index.js:29892:12)
    at async Proxy._request (/Users/antoine/Documents/www/karpeo/frontend/node_modules/@prisma/client/runtime/index.js:30864:16)
    at async Promise.all (index 0)
    at async Object.handler (/Users/antoine/Documents/www/karpeo/frontend/cypress.config.localhost.js:76:9)
```

I've checked, despite the error being "Invalid `prisma.$executeRaw()` invocation:", the error is caused by the create call which, I assume, calls executeRaw eventually.

When executing Sql against the database directly, I can get the following SQL to work and is what I trying to achieve with prisma:

```sql
BEGIN TRAN
SET IDENTITY_INSERT Customer ON;
INSERT INTO Customer (id, ...)
VALUES (6, ...)
SET IDENTITY_INSERT Customer OFF;
COMMIT
```

And I missing something very obvious? When searching for `IDENTITY_INSERT + prisma` on google, I literally get 0 results. Same thing here and in github issues. I assume SqlServer isn't being used very often with prisma.

Thanks 👍

**Comment by austincrim** [comment, marked as answer]
Hey @CapitaineToinon 👋,

Could you open a [bug report](https://github.com/prisma/prisma/issues/new?assignees=&labels=kind%2Fbug&template=bug_report.yml) with this info? I'm also not seeing any relevant history with Prisma and `IDENTITY_INSERT` so it's possible we haven't tested this workflow.

**Reply by CapitaineToinon** [reply]
Thanks for the quick reply, @austincrim! I created the issue #15305
